core/src/test/java/edu/wpi/grip/core/sources/CameraSourceTest.java

```diff
@@ -0,0 +1,154 @@
+        }
+
+        public void setShouldThrowAtStart(boolean shouldThrowAtStart){
+            this.shouldThrowAtStart = shouldThrowAtStart;
+        }
+    }
+
+    class MockFrameGrabberFactory implements CameraSource.FrameGrabberFactory {
+        private MockFrameGrabber frameGrabber = new MockFrameGrabber();
+        @Override
+        public FrameGrabber create(int deviceNumber) {
+            return frameGrabber;
+        }
+
+        @Override
+        public FrameGrabber create(String addressProperty) throws MalformedURLException {
+            return frameGrabber;
+        }
+    }
+
+    @Before
+    public void setUp() throws Exception {
+        this.injector = Guice.createInjector(new GRIPCoreModule());
+        this.cameraSourceFactory = injector.getInstance(CameraSource.Factory.class);
+
+        this.eventBus = new EventBus();
+        class UnhandledExceptionWitness {
+            @Subscribe
+            public void onUnexpectedThrowableEvent(UnexpectedThrowableEvent event) {
+                System.err.println(event.getMessage());
+                event.getThrowable().printStackTrace();
+            }
+        }
+        this.eventBus.register(new UnhandledExceptionWitness());
+        this.mockFrameGrabberFactory = new MockFrameGrabberFactory();
+        this.cameraSourceWithMockGrabber = new CameraSource(eventBus, mockFrameGrabberFactory, 0);
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        mockFrameGrabberFactory.frameGrabber.release();
+    }
+
+    @Test(expected = IOException.class)
+    public void testInvalidURLThrowsIOException() throws Exception {
+        cameraSourceFactory.create("Not a URL at all!");
+        fail("This test should have thrown an exception.");
+    }
+
+    @Test
+    public void testCanStopAndStart() throws Exception {
+        cameraSourceWithMockGrabber.start();
+        assertTrue("The camera source was not started after calling start", cameraSourceWithMockGrabber.isStarted());
+        cameraSourceWithMockGrabber.stop();
+        assertFalse("The camera was not stopped after calling stop", cameraSourceWithMockGrabber.isStarted());
+    }
+
+    @Test(expected = IOException.class)
+    public void testStartRethrowsIfFailure() throws Exception {
+        mockFrameGrabberFactory.frameGrabber.setShouldThrowAtStart(true);
+        cameraSourceWithMockGrabber.start();
+        fail("Should have thrown an IOException");
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void testStartingTwiceShouldThrowIllegalState() throws Exception {
+        try {
+            try {
+                cameraSourceWithMockGrabber.start();
+            } catch(RuntimeException e) {
+                fail("This should not have failed");
+            }
+            cameraSourceWithMockGrabber.start();
+        } finally {
+            cameraSourceWithMockGrabber.stop();
+        }
+        fail("The test should have failed with an IllegalStateException");
+    }
+
```

Comment:
Add a test that calling stop where stop will throw an exception still stops the thread.